src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/Streaming/DataStreamWriterTests.cs

```diff
@@ -71,44 +71,45 @@ public void TestSignaturesV2_3_X()
         [SkipIfSparkVersionIsLessThan(Versions.V2_4_0)]
         public void TestForeachBatch()
         {
-            // Temporary folder to put our test stream input.
-            using var srcTempDirectory = new TemporaryDirectory();
+            var memoryStream = new MemoryStream<int>(_spark);
+
             // Temporary folder to write ForeachBatch output.
             using var dstTempDirectory = new TemporaryDirectory();
 
             Func<Column, Column> outerUdf = Udf<int, int>(i => i + 100);
 
-            // id column: [0, 1, ..., 9]
-            WriteCsv(0, 10, Path.Combine(srcTempDirectory.Path, "input1.csv"));
-
-            DataStreamWriter dsw = _spark
-                .ReadStream()
-                .Schema("id INT")
-                .Csv(srcTempDirectory.Path)
+            DataStreamWriter dsw = memoryStream
+                .ToDF()
                 .WriteStream()
                 .ForeachBatch((df, id) =>
                 {
                     Func<Column, Column> innerUdf = Udf<int, int>(i => i + 200);
-                    df.Select(outerUdf(innerUdf(Col("id"))))
+                    df.Select(outerUdf(innerUdf(Col("value"))))
                         .Write()
                         .Csv(Path.Combine(dstTempDirectory.Path, id.ToString()));
                 });
 
             StreamingQuery sq = dsw.Start();
 
-            // Process until all available data in the source has been processed and committed
-            // to the ForeachBatch sink. 
-            sq.ProcessAllAvailable();
+            TimeSpan streamingTimeout = TimeSpan.FromSeconds(30);
+            try
+            {
+                memoryStream.AddData(Enumerable.Range(0, 10).ToArray());
+
+                // Process until all available data in the source has been processed and committed
+                // to the ForeachBatch sink.
+                FailAfter(streamingTimeout, () => sq.ProcessAllAvailable());
```

Comment:
Do we also expect the timeout with this approach? If so, what's the benefit of doing this over the previous approach?

Reply:
We can determine if it's this test that is hanging.